### PR TITLE
(BKR-394) added support for 3.8/AIO mixed installations

### PR DESCRIPTION
**Pull request by kevpl**

### CODE REVIEW

File: lib/beaker/dsl/install_utils/pe_utils.rb [hunk 1]
```diff
@@ -512,7 +516,11 @@ def create_agent_specified_arrays(hosts)
           non_agent_only_roles = %w(master database dashboard console frictionless)
           hosts.each do |host|
             if host['roles'].none? {|role| non_agent_only_roles.include?(role) }
-              hosts_agent_only << host
+              if host[:pe_ver] && version_is_less(host[:pe_ver], '4.0')
```
Comment (colinPL) [inline lib/beaker/dsl/install_utils/pe_utils.rb]: I see this check twice - here and again in `roles.rb`. Can this be moved to a method? `is_aio_capable?` or something?

File: lib/beaker/dsl/roles.rb [hunk 2]
```diff
@@ -115,7 +115,11 @@ def not_controller(host)
       #       puts "this host is ONLY an agent!"
       #     end
       def agent_only(host)
-        host['roles'].length == 1 && host['roles'].include?('agent')
+        if host[:pe_ver] && version_is_less(host[:pe_ver], '4.0')
+          false
```
Comment (doug-rosser) [inline lib/beaker/dsl/roles.rb]: the test "agent_only" becomes a bit ambiguous here, because there's a small chance someone might consider a "legacy agent" as an agent. This is admittedly a corner case!

Comment (kevpl) [inline lib/beaker/dsl/roles.rb]: it is an agent, it's just not "agent_only", which is a special kind of agent.

that does seem confusing, however.  Can we think of a better name for it?  That would be a much bigger diff, but I'm open to it if something more specific strikes.  `aio_agent?`, perhaps?

**Comment by doug-rosser** [comment]
:+1:  I've already seen @johnduarte report success manually testing installs

File: lib/beaker/dsl/roles.rb (hunk 2 above)
Comment (anodelman) [inline lib/beaker/dsl/roles.rb]: Yeah, I don't like mixing this logic into agent_only, which is only for indicating if a host has the single role 'agent'.

A separate method is in order for what you are trying to determine.

**Comment by kevpl** [comment]
I manually stopped that jenkins run.  a new one has already kicked off from my latest push

**Comment by kevpl** [comment]
@puppetlabs-jenkins retest this please

**Comment by anodelman** [comment]
@kevpl you ready to have this merged?  Did you diagnose the ride along issue?

**Comment by kevpl** [comment]
@anodelman once it goes through this time, then I'll be ready for merge.

The issue before was from an old assumption.  In AIO, the host property `privatebindir` was introduced because the dev teams' were splitting up where the vendored binaries and the rest go.  When we set that up, the pe version was set to be the same as the new AIO version, since it had never existed before.  It should have been set to the old `puppetbindir` path, so that you could always use it to find ruby and the vendored bins, however.  The path change fixes that.

**Comment by anodelman** [comment]
retest this please

File: lib/beaker/dsl/install_utils/pe_defaults.rb [hunk 3]
```diff
@@ -32,7 +32,7 @@ module PEDefaults
             'puppetbin'        => '/opt/puppet/bin/puppet',
             'puppetbindir'     => '/opt/puppet/bin',
             'puppetsbindir'    => '/opt/puppet/sbin',
-            'privatebindir'    => '/opt/puppetlabs/puppet/bin',
+            'privatebindir'    => '/opt/puppet/bin',
```
Comment (johnduarte) [inline lib/beaker/dsl/install_utils/pe_defaults.rb]: This winds up flipping the desired values. Looks like I gave you a bum steer @kevpl

Comment (kevpl) [inline lib/beaker/dsl/install_utils/pe_defaults.rb]: after talking with John, we realized that this is the correct path, but that other logic was causing the wrong hosts to be set with their paths, so we've updated that instead of this.

**Comment by anodelman** [comment]
Green on classifier tests.

**Comment by anodelman** [comment]
retest this please
